clang-tools-extra/test/clang-tidy/checkers/bugprone/reverse-null.cpp

```diff
@@ -0,0 +1,207 @@
+// RUN: %check_clang_tidy %s bugprone-experimental-reverse-null %t
+
+struct S {
+  int a;
+};
+
+int warning_deref(int *p) {
+  *p = 42;
+
+  if (p) {
+    // CHECK-MESSAGES: :[[@LINE-1]]:7: warning: pointer value is checked despite dereferencing it earlier
+    *p += 20;
+    return *p;
+  } else {
+    return 0;
+  }
+}
+
+int warning_member(S *q) {
+  q->a = 42;
+
+  if (q) {
+    // CHECK-MESSAGES: :[[@LINE-1]]:7: warning: pointer value is checked despite dereferencing it earlier
+    q->a += 20;
+    return q->a;
+  } else {
+    return 0;
+  }
+}
+
+int negative_warning(int *p) {
+  *p = 42;
+
+  if (!p) {
+    // CHECK-MESSAGES: :[[@LINE-1]]:8: warning: pointer value is checked despite dereferencing it earlier
+    return 0;
+  } else {
+    *p += 20;
+    return *p;
+  }
+}
+
+int no_warning(int *p, bool b) {
+  if (b) {
+    *p = 42;
+  }
+
+  if (p) {
+    // CHECK-MESSAGES-NOT: :[[@LINE-1]]:7: warning: pointer value is checked despite dereferencing it earlier 
+    *p += 20;
+    return *p;
+  } else {
+    return 0;
+  }
+}
+
+int propagate_condition(int *p, bool b) {
+  if (b) {
+    *p = 20;
+  }
+
+  (void)p;
+
+  if (b) {
+    if (p) {
+    // CHECK-MESSAGES: :[[@LINE-1]]:9: warning: pointer value is checked despite dereferencing it earlier 
+      *p += 15;
+      return *p;
+    } else {
+      return 0;
+    }
+  } else {
+    return 1;
+  }
```

Comment:
Impossible branches are filtered out using the flow condition, but in this case, even the CFG doesn't have an edge to after the if, since all branches return.